### PR TITLE
[JENKINS-60266, JEP-223] Add Jenkins.MANAGE permission and allow some item of /manage with it

### PR DESCRIPTION
**Pull request by EstherAF**
See [JENKINS-60266](https://issues.jenkins-ci.org/browse/JENKINS-60266).
But also the [JEP-223](https://github.com/jenkinsci/jep/tree/master/jep/223)
And also the topic in the dev list: https://groups.google.com/forum/#!topic/jenkinsci-dev/lThSifB8G4k
This PR is a fork of the #4374, in order to keep the history cleaner.

This is an *experimental feature*, the new Jenkins.MANAGE permission is disabled by default, can be enabled with the system property "jenkins.security.ManagePermission"

### Proposed changelog entries
* Entry 1: Add the Overall/Manage permission, that allows a user to partially configure Jenkins without the Overall/Administer permission. This is an experimental feature, disabled by default, that can be enabled with "jenkins.security.ManagePermission" system property.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@oleg-nenashev @daniel-beck @Wadeck 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Comment by EstherAF** [comment]
When the permission is disabled, some Access Denied messages are showing "user is missing the Overall/Manage permission". 
This is also happening with the `Job/ExtendedRead` permission.
There is a PR to fix this (#4482)

**Comment by daniel-beck** [comment]
Requested reviews from commenters/reviewers on the original PR.

File: core/src/main/java/jenkins/management/CliLink.java [hunk 1]
```diff
@@ -52,4 +56,9 @@ public String getDescription() {
     public String getUrlName() {
         return "cli";
     }
+
+    @Override
+    public Permission getRequiredPermission() {
+        return Jenkins.MANAGE;
+    }
```
Comment (jtnord) [inline core/src/main/java/jenkins/management/CliLink.java]: this always struck me as odd.  the CLI is available for everyone to use and each CLI should have its own authz on who can use it inside the implementation - but users have no way to actually discover these. 

this is not related to this PR - it has not made this worse, just a note.

File: core/src/main/java/jenkins/management/SystemInfoLink.java [hunk 2]
```diff
@@ -52,4 +56,9 @@ public String getDescription() {
     public String getUrlName() {
         return "systemInfo";
     }
+
+    @Override
+    public Permission getRequiredPermission() {
+        return Jenkins.MANAGE;
```
Comment (jtnord) [inline core/src/main/java/jenkins/management/SystemInfoLink.java]: depending on how Jenkins is run - could this expose passwords for TLS certs which are not configurable in the UI?

File: core/src/main/java/jenkins/model/GlobalConfiguration.java [hunk 3]
```diff
@@ -75,4 +82,36 @@ public boolean configure(StaplerRequest req, JSONObject json) throws FormExcepti
...
+                    return Functions.hasPermission(((GlobalConfiguration) descriptor).getPermission());
+                } else {
+                    return true;
+                }
+            } catch (AccessDeniedException e) {
+                return false;
+            } catch (ServletException | IOException e) {
+                e.printStackTrace();
+                return true;
+            }
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/GlobalConfiguration.java]: can be simplified (I think) to the following:
```java
               if (descriptor instanceof GlobalConfiguration) {
                    return Jenkins.get().hasPermission(((GlobalConfiguration) descriptor).getPermission());
                }
                return true;
```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -2189,14 +2193,18 @@ public AdministrativeMonitor getAdministrativeMonitor(String id) {
...
     public List<AdministrativeMonitor> getActiveAdministrativeMonitors() {
-        return administrativeMonitors.stream().filter(m -> {
-            try {
-                return m.isEnabled() && m.isActivated();
-            } catch (Throwable x) {
-                LOGGER.log(Level.WARNING, null, x);
-                return false;
-            }
-        }).collect(Collectors.toList());
+        if (Jenkins.get().hasPermission(ADMINISTER)){
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: but some `AdministrativeMonitors` are accessible with MANAGE (in this PR - see OldDataMonitor for example?)

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 5]
```diff
@@ -4353,7 +4361,7 @@ protected RestartCause() {
     @CLIMethod(name="shutdown")
     @RequirePOST
     public void doExit( StaplerRequest req, StaplerResponse rsp ) throws IOException {
-        checkPermission(ADMINISTER);
+        checkPermission(MANAGE);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: can the MANAGER user recover from this?  (could they restart it).  its not a security issue theres no leak but just wondering if they should be allowed to do this (what would the use case be)

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 6]
```diff
@@ -4386,7 +4394,7 @@ public void run() {
     @CLIMethod(name="safe-shutdown")
     @RequirePOST
     public HttpResponse doSafeExit(StaplerRequest req) throws IOException {
-        checkPermission(ADMINISTER);
+        checkPermission(MANAGE);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: can the MANAGER user recover from this? (could they restart it). its not a security issue theres no leak but just wondering if they should be allowed to do this (what would the use case be)

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 7]
```diff
@@ -5226,7 +5234,24 @@ private static void computeVersion(ServletContext context) {
...
     public static final Permission ADMINISTER = Permission.HUDSON_ADMINISTER;
+
+    /**
+     * Allows non-privilege escalating configuration permission for a Jenkins instance.  Actions which could result
+     * in a privilege  escalation (such as RUN_SCRIPTS) require explicit ADMINISTER permission.
+     *
+     * As an experimental feature, making the manage permission able to be disabled by default (keep as ADMINISTER), can
+     * be enabled with "jenkins.security.ManagePermission" system property.
+     */
+    public static final Permission MANAGE = new Permission(PERMISSIONS, "Manage",
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: where we not trying to mark this as `@Beta` for the moment?

File: core/src/main/resources/hudson/model/Messages.properties [hunk 8]
```diff
@@ -163,6 +163,10 @@ Hudson.AdministerPermission.Description=\
   This permission grants the ability to make system-wide configuration changes, \
   as well as perform highly sensitive operations that amounts to full local system access \
   (within the scope granted by the underlying OS.)
+Hudson.ManageJenkins.Description=\
+  This permission grants the ability to make non Permission escalating system \
+  configuration changes. \
+  System configuration which would allow the execution of arbitrary commands or code on the master requires Administer.
```
Comment (jtnord) [inline core/src/main/resources/hudson/model/Messages.properties]: @HCloward I think there should be some commas in that sentence somewhere but I'm not sure whre.

File: core/src/main/resources/hudson/slaves/SlaveComputer/threadDump.jelly [hunk 9]
```diff
@@ -33,13 +33,13 @@ THE SOFTWARE.
     <st:include page="sidepanel.jelly" />
 
     <l:main-panel>
-      <l:isAdmin>
+      <j:if test="${h.hasPermission(app.MANAGE)}">
```
Comment (jtnord) [inline core/src/main/resources/hudson/slaves/SlaveComputer/threadDump.jelly]: unsure about this one.  will go with the majority.

File: core/src/main/resources/jenkins/management/AdministrativeMonitorsConfiguration/config.groovy [hunk 10]
```diff
@@ -25,28 +25,31 @@
...
-                        f.checkbox(name: "administrativeMonitor",
-                                title: am.displayName,
-                                checked: am.enabled,
-                                json: am.id)
-                    }
-                    tr() {
-                        td(colspan: "2")
-                        td(class: "setting-description") {
-                            st.include(from: am, page: "description", optional: true)
+if (Jenkins.get().hasPermission(Jenkins.ADMINISTER)) {
```
Comment (jtnord) [inline core/src/main/resources/jenkins/management/AdministrativeMonitorsConfiguration/config.groovy]: see previous comment - some administrative monitors have their permission set to `MANAGE` - do we have a consistent UX here?

File: test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * This tests can be merged into {@link InstallPluginCommandTest} when the system property
+ * {@code jenkins.security.ManagePermission} is no longer supported
+ */
+public class InstallPluginCommandWithManageTest {
+
+    static {
+        // this happens before the Jenkins static fields are loaded
+        System.setProperty("jenkins.security.ManagePermission", "true");
```
Comment (jtnord) [inline test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java]: Surefire is reusing forks - so this needs to be in a `@BeforeClass` method and cleared in an `@AfterClass` method IIUC

File: test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    private void setupUpdateCenter() {
+        try {
+            r.jenkins.getUpdateCenter().getSite(UpdateCenter.ID_DEFAULT).updateDirectlyNow(false);
```
Comment (jtnord) [inline test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java]: why are you ignoring the signature check?

File: test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        } catch (Exception x) {
+            assumeNoException(x);
+        }
+        InetSocketAddress address = new InetSocketAddress("updates.jenkins-ci.org", 80);
+        assumeFalse("Unable to resolve updates.jenkins-ci.org. Skip test.", address.isUnresolved());
```
Comment (jtnord) [inline test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java]: `updateDirectlyNow` returns a `FormValidation` - do you need to do these checks here, or can you just check that result?

File: test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+ * with this property activated.
+ */
+public class JenkinsManagePermissionTest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    static {
+        // this happens before the Jenkins static fields are loaded
+        System.setProperty("jenkins.security.ManagePermission", "true");
```
Comment (jtnord) [inline test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java]: see previous comment about surefire reusing forks and SystemProperties

File: test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+    @Test
+    public void globalConfigAllowedWithManagePermission() throws Exception {
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(new MockAuthorizationStrategy()
+                                                   .grant(Jenkins.MANAGE, Jenkins.READ).everywhere().toEveryone());
+
+        HtmlForm form = j.createWebClient().goTo("configure").getFormByName("config");
+        HtmlPage updated = j.submit(form);
+        assertEquals("User with Jenkins.MANAGE permission should be able to update global configuration",
+                     200, updated.getWebResponse().getStatusCode());
```
Comment (jtnord) [inline test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java]: NIT: if this returns something other than `200` when run on a CI system will there be enough information to diagnose the test failure?

if this is done in multiple places, would a hamcrest matcher provide more diagnostics?

`assertThat("User with Jenkins.MANAGE permission should be able to update global configuration", updated.getWebResponse(), hasStatusCode(200))`

and describeMismatch could say the error code and the contents.

**Review by jtnord — COMMENTED** [review]
the [`VisibilityFilter`](https://github.com/jenkinsci/jenkins/pull/4483/files#diff-841df66889e3c3214c68de134dfe05beR1074) change looks like a possible regression.
other than that just questions or comments - looks good.

File: core/src/main/java/hudson/model/ManagementLink.java [hunk 16]
```diff
@@ -112,10 +112,10 @@ public boolean getRequiresConfirmation() {
     }
 
     /**
-     * @return permission required for user to access this management link, in addition to {@link Jenkins#ADMINISTER}
+     * @return permission required for user to access this management link, default is {@link Jenkins#ADMINISTER}
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/ManagementLink.java]: AFAICT it is a breaking change in API. The default behavior changes. I doubt we care much after the removal of dangerous permissions, but it would be great to have an explicit usage abalysis

File: core/src/main/java/hudson/PluginWrapper.java [hunk 17]
```diff
@@ -439,7 +439,7 @@ public String getDisplayName() {
     }
 
     public Api getApi() {
-        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+        if (!Jenkins.get().hasPermission(Jenkins.ADMINISTER)) return null;
```
[the remark this answers is not in the document]
Comment (daniel-beck) [reply core/src/main/java/hudson/PluginWrapper.java]: The problem here is that `thirdPartyLicenses` (linked from `/about`) is served by `PluginWrapper`, but `layout.jelly` calls `getApi()` to determine whether to show a link in the footer. Without this change, that would show an exception stack trace at least in the log, possibly on the page, for Overall/Manage users.

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java [hunk 18]
```diff
@@ -458,5 +459,15 @@ public String getDescription() {
...
+
+        @Override
+        public Permission getRequiredPermission() {
+            return Jenkins.MANAGE;
+        }
+    }
+
+    @Override
+    public Permission getRequiredPermission() {
+        return Jenkins.MANAGE;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/diagnosis/OldDataMonitor.java]: This change (and the monitor) is undesirable, since Manage does not imply global Item/Read, or even Item/Discover, so this might list items the user is not entitled to see.

Or it needs to be documented _very_ carefully.

File: core/src/main/java/jenkins/management/CliLink.java (hunk 1 above)
Comment (daniel-beck) [reply core/src/main/java/jenkins/management/CliLink.java]: Right, basically just limiting the discovery of the CLI page. That said, this change is correctly not hiding the link from Manage-rs.

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java (hunk 18 above)
Comment (mikecirioli) [reply core/src/main/java/hudson/diagnosis/OldDataMonitor.java]: I am leaning toward moving this back to requiring `Jenkins.ADMINISTER`

**Comment by mikecirioli** [comment]
@timja if you have some time, we would appreciate your review on this PR

File: core/src/main/java/hudson/AboutJenkins.java [hunk 19]
```diff
@@ -38,4 +42,9 @@ public URL getLicensesURL() {
         return AboutJenkins.class.getResource("/META-INF/licenses.xml");
     }
 
+    @Override
+    public Permission getRequiredPermission() {
+        return Jenkins.MANAGE;
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/AboutJenkins.java]: Should it be replaced by read-only access in JEP-224?

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java (hunk 18 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/diagnosis/OldDataMonitor.java]: +1 for reverting

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 20]
```diff
@@ -3752,7 +3760,7 @@ public Object getDynamic(String token) {
     public synchronized void doConfigSubmit( StaplerRequest req, StaplerResponse rsp ) throws IOException, ServletException, FormException {
         BulkChange bc = new BulkChange(this);
         try {
-            checkPermission(ADMINISTER);
+            checkPermission(MANAGE);
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: Taking how the code below is implemented, it is likely to expose some Admin-only configuration sections to users with `Manage` permissions. Hiding controls from frontend is fine, but IIUC the manually prepared from submission would pass here.

**Review by oleg-nenashev — DISMISSED** [review]
I would like to understand how the current implementation protects Unclassified Global configuration items from being modified by handcrafted form submissions by users with `Manage` permissions. If I understand the code correctly, it would require all implementations to explicitly check permissions, similarly to how `GlobalNodePropertiesConfiguration` is implemented in this pull request. So all unclassified GlobalConfiguration items in plugins are now accessible with `Manage` permissions. 

There are some security aspects being managed as unclassified GlobalConfigurations, e.g. Support Bundle Anonymization in the Support Core plugin. According to JEP-223 such security-related controls should be restricted to admins by default.

Requesting changes to get clarification and, ideally, a direct unit test for the suggested case or its equivalent

File: core/src/main/java/hudson/AboutJenkins.java (hunk 19 above)
Comment (EstherAF) [reply core/src/main/java/hudson/AboutJenkins.java]: I think you are right, as this view can not be used to modify anything. It can be changed to read-only when #4149 is merged, or as part of it maybe.

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java (hunk 18 above)
Comment (mikecirioli) [reply core/src/main/java/hudson/diagnosis/OldDataMonitor.java]: reverted back to `Jenkins.ADMINISTER`

File: test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java (hunk 11 above)
Comment (mikecirioli) [reply test/src/test/java/hudson/cli/InstallPluginCommandWithManageTest.java]: fixed

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (mikecirioli) [reply core/src/main/java/hudson/diagnosis/OldDataMonitor.java]: Permission check has been reverted back to `Jenkins.ADMINISTER`

File: core/src/main/java/hudson/model/ManagementLink.java (hunk 16 above)
Comment (mikecirioli) [reply core/src/main/java/hudson/model/ManagementLink.java]: This is actually just a change correcting a mistake in the javadoc.  The method itself originally returned `null` (I presume to help ensure subclasses actually provided a specific permission), and a access to a `ManagementLink` is controlled by a single permission.

File: core/src/main/java/jenkins/management/SystemInfoLink.java (hunk 2 above)
Comment (mikecirioli) [reply core/src/main/java/jenkins/management/SystemInfoLink.java]: Can you provide an example of what you are thinking about?

File: core/src/main/java/hudson/model/ManagementLink.java (hunk 16 above)
Comment (aHenryJard) [reply core/src/main/java/hudson/model/ManagementLink.java]: I am going to look for usages in plugins.
A least in Jenkins core scope, this is not a breaking change as this "getRequiredPermission()" is only used to show or not show an item in the "/manage" page. And that page is right now only visible for ADMINISTER. so there is no change there for a user with overall/READ for example.

Comment (aHenryJard) [reply core/src/main/java/hudson/model/ManagementLink.java]: I will try to make it clearer on the JEP also.

Comment (aHenryJard) [reply core/src/main/java/hudson/model/ManagementLink.java]: @oleg-nenashev do you have a nice example of usage analysis ? I don't know what it could looks likes.

File: core/src/main/java/hudson/AboutJenkins.java (hunk 19 above)
Comment (EstherAF) [reply core/src/main/java/hudson/AboutJenkins.java]: @timja we need this because the change in `ManagementLink#getRequiredPermission`.

If possible, I think would be better to continue the conversation there so it is easier to follow :smile: 

https://github.com/jenkinsci/jenkins/pull/4483/files/78561c11542e692c71442264b2f94ea7e04a0263#diff-97a16a57b311b64546f011915a4ab1ba

File: core/src/main/java/jenkins/management/AdministrativeMonitorsConfiguration.java [hunk 21]
```diff
@@ -42,6 +44,9 @@
 public class AdministrativeMonitorsConfiguration extends GlobalConfiguration {
     @Override
     public boolean configure(StaplerRequest req, JSONObject json) throws FormException {
+        if (!Jenkins.get().hasPermission(Jenkins.ADMINISTER)){
```
[the remark this answers is not in the document]
Comment (EstherAF) [reply core/src/main/java/jenkins/management/AdministrativeMonitorsConfiguration.java]: you were right, the other Administrative Monitors that were `MANAGE` were errors. We've reverted them to `ADMINISTER`

**Comment by EstherAF** [comment]
> I would like to understand how the current implementation protects Unclassified Global configuration items from being modified by handcrafted form submissions by users with `Manage` permissions. If I understand the code correctly, it would require all implementations to explicitly check permissions, similarly to how `GlobalNodePropertiesConfiguration` is implemented in this pull request. So all unclassified GlobalConfiguration items in plugins are now accessible with `Manage` permissions.
>
> There are some security aspects being managed as unclassified GlobalConfigurations, e.g. Support Bundle Anonymization in the Support Core plugin. According to JEP-223 such security-related controls should be restricted to admins by default.

@oleg-nenashev 

You are right, all implementations will have to check the permissions manually. 

We are open to solve this in the near future in a posterior PR, making all Unclassified Global configuration items accessible only by Administer by default. As there are several different ways of achieving it, and this PR is already making several different changes, we would like to avoid deciding how to do it as part of this PR. As this would be released as an experimental feature disabled by default, the existent behaviour related to Unclassified Global configuration items won't change unless users the feature.

**Comment by oleg-nenashev** [comment]
@EstherAF would it be possible to document the current limitations in the JEP? I am fine if the experimental feature does not behave as is in very beginning, but there should be explicitly documented (preferably with a list of affected plugins)

**Comment by mikecirioli** [comment]
> @EstherAF would it be possible to document the current limitations in the JEP? I am fine if the experimental feature does not behave as is in very beginning, but there should be explicitly documented (preferably with a list of affected plugins)

We are working on exactly that documentation @oleg-nenashev (although pushing JEP updates has taken a slightly lower priority this week as we work to address the code feedback in this PR).  We are committed to including this information in the JEP.

File: test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java (hunk 15 above)
Comment (mikecirioli) [reply test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java]: resolved

File: test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java (hunk 14 above)
Comment (mikecirioli) [reply test/src/test/java/jenkins/model/JenkinsManagePermissionTest.java]: resolved

File: core/src/main/java/hudson/Functions.java [hunk 22]
```diff
@@ -1071,7 +1071,7 @@ public static String getFooterURL() {
         List<Descriptor> answer = new ArrayList<>(r.size());
         for (Tag d : r) answer.add(d.d);
 
-        return DescriptorVisibilityFilter.apply(Jenkins.get(),answer);
+        return GlobalConfiguration.GlobalConfigHiddenByPermissionFilter.apply(Jenkins.get(),answer);
```
[the remark this answers is not in the document]
Comment (mikecirioli) [reply core/src/main/java/hudson/Functions.java]: resolved by pr https://github.com/EstherAF/jenkins/pull/4

**Comment by daniel-beck** [comment]
> I would like to understand how the current implementation protects Unclassified Global configuration items from being modified by handcrafted form submissions by users with Manage permissions.

This is done pretty nicely, as long as we don't consider the System Read proposal. Read along:

3771 calls `Functions#getSortedDescriptorsForGlobalConfigUnclassified`:

https://github.com/jenkinsci/jenkins/blob/4084a712b60dccd8bbf16a3bc359bce3b289e217/core/src/main/java/jenkins/model/Jenkins.java#L3760-L3771

1097 calls `#getSortedDescriptorsForGlobalConfig`:

https://github.com/jenkinsci/jenkins/blob/2fcd83186686cd123ca602af23eb0a1cc2f93a01/core/src/main/java/hudson/Functions.java#L1096-L1102
https://github.com/jenkinsci/jenkins/blob/2fcd83186686cd123ca602af23eb0a1cc2f93a01/core/src/main/java/hudson/Functions.java#L1074

So just like the form [hides descriptors for which the user does not have permission](https://github.com/jenkinsci/jenkins/blob/2fcd83186686cd123ca602af23eb0a1cc2f93a01/core/src/main/resources/jenkins/model/Jenkins/configure.jelly#L48), the server-side form submission handling code does not process data related to descriptors you have no permission to. This is all accomplished by the new:

https://github.com/jenkinsci/jenkins/blob/4084a712b60dccd8bbf16a3bc359bce3b289e217/core/src/main/java/jenkins/model/GlobalConfiguration.java#L102-L112

and the default is safe:

https://github.com/jenkinsci/jenkins/blob/4084a712b60dccd8bbf16a3bc359bce3b289e217/core/src/main/java/jenkins/model/GlobalConfiguration.java#L92-L94

File: core/src/main/java/hudson/model/ManagementLink.java (hunk 16 above)
Comment (aHenryJard) [reply core/src/main/java/hudson/model/ManagementLink.java]: The JEP related description is here https://github.com/jenkinsci/jep/pull/269

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 7 above)
Comment (EstherAF) [reply core/src/main/java/jenkins/model/Jenkins.java]: The interface for this permission is not expected to change. Why should we consider it Beta?

Comment (daniel-beck) [reply core/src/main/java/jenkins/model/Jenkins.java]: This is still intended to be delivered as an experimental feature, right?

If so, makes sense to require a minor hurdle to use it. Strictly speaking, all the `#getRequiredPermission()` and such should be annotated, but this one's probably more straightforward.

File: core/src/main/java/hudson/model/ManagementLink.java (hunk 16 above)
Comment (Wadeck) [reply core/src/main/java/hudson/model/ManagementLink.java]: Hiding this change behind the feature flag could be useful

**Review by Wadeck — COMMENTED** [review]
Only reviewed down to ManagementLink at the moment, testing the cases manually and looking for the usage in ecosystem take time :(

File: core/src/main/resources/hudson/security/HudsonPrivateSecurityRealm/sidepanel.jelly [hunk 23]
```diff
@@ -28,7 +28,7 @@ THE SOFTWARE.
   <l:side-panel>
     <l:tasks>
       <l:task href="${rootURL}/" icon="icon-up icon-md" title="${%Back to Dashboard}"/>
-      <l:task href="${rootURL}/manage" icon="icon-gear2 icon-md" permission="${app.ADMINISTER}" title="${%Manage Jenkins}"/>
+      <l:task href="${rootURL}/manage" icon="icon-gear2 icon-md" permission="${app.MANAGE}" title="${%Manage Jenkins}"/>
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/security/HudsonPrivateSecurityRealm/sidepanel.jelly]: This entire view isn't ever used if you don't have ADMINISTER.

Comment (Wadeck) [reply core/src/main/resources/hudson/security/HudsonPrivateSecurityRealm/sidepanel.jelly]: => <jenkins-url>/securityRealm/success is reachable by just Overall/Read
So this change makes sense :)

File: core/src/main/java/hudson/ProxyConfiguration.java [hunk 24]
```diff
@@ -428,7 +428,7 @@ public FormValidation doValidateProxy(
                 @QueryParameter("userName") String userName, @QueryParameter("secretPassword") Secret password,
                 @QueryParameter("noProxyHost") String noProxyHost) {
 
-            Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+            Jenkins.get().checkPermission(Jenkins.MANAGE);
```
[the remark this answers is not in the document]
Comment (mikecirioli) [reply core/src/main/java/hudson/ProxyConfiguration.java]: This could be safely reverted.  We have discussed the idea of moving the proxy configuration to a more sensible place (would be done in a unrelated PR) since the proxy is not strictly used  for just plugin installation.

File: core/src/main/java/hudson/model/AdministrativeMonitor.java [hunk 25]
```diff
@@ -154,11 +155,18 @@ public void doDisable(StaplerRequest req, StaplerResponse rsp) throws IOExceptio
...
+        return Jenkins.ADMINISTER;
+    }
+
+    /**
+     * Check required permission for any operation in here.
      */
     @Restricted(NoExternalUse.class)
     public Object getTarget() {
-        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+        Jenkins.get().checkPermission(getRequiredPermission());
```
[the remark this answers is not in the document]
Comment (mikecirioli) [reply core/src/main/java/hudson/model/AdministrativeMonitor.java]: I do not believe so, at least not currently.  Due to the potential risk we had decided to leave admin monitors as strictly `Jenkins.ADMINISTER` for now.  Possibly we overlooked this change from an earlier commit.

**Comment by batmat** [comment]
@EstherAF FYI the PR is conflicted now.

**Comment by Wadeck** [comment]
At the moment, this PR should not be merged, see the simpler version: https://github.com/jenkinsci/jenkins/pull/4483

The changes that are done in this PR but not yet included in #4483 will be provided as a separate following-up PR. This one represents the full vision of the desired changes.

Putting on-hold for the moment.

File: core/src/main/resources/hudson/model/Computer/sidepanel.jelly [hunk 26]
```diff
@@ -37,7 +37,8 @@ THE SOFTWARE.
       <l:task href="${rootURL}/${it.url}builds" icon="icon-notepad icon-md" title="${%Build History}"/>
       <l:task href="${rootURL}/${it.url}load-statistics" icon="icon-monitor icon-md" title="${%Load Statistics}"/>
       <j:if test="${it.channel!=null}">
-        <l:task href="${rootURL}/${it.url}script" icon="icon-terminal icon-md" permission="${app.RUN_SCRIPTS}" title="${%Script Console}"/>
+        <l:task href="${rootURL}/${it.url}script" icon="icon-terminal icon-md" permission="${app.ADMINISTER}"
```
Comment (varyvol) [inline core/src/main/resources/hudson/model/Computer/sidepanel.jelly]: Why does this one need to be changed?

**Comment by daniel-beck** [comment]
It makes no sense for this to remain open as the core functionality was merged. Ideally, individual self-contained PRs opening up specific features to Manage users should be submitted (like Tim is doing for System Read).
